File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/property/ConstrainedSize.java
/* Copyright Airship and Contributors */

package com.urbanairship.android.layout.property;

import com.urbanairship.android.layout.util.PercentUtils;
import com.urbanairship.json.JsonException;
import com.urbanairship.json.JsonMap;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import static com.urbanairship.android.layout.property.ConstrainedSize.ConstrainedDimensionType.ABSOLUTE;
import static com.urbanairship.android.layout.property.ConstrainedSize.ConstrainedDimensionType.PERCENT;

public class ConstrainedSize extends Size {
    @Nullable
    private final ConstrainedDimension minWidth;
    @Nullable
    private final ConstrainedDimension minHeight;
    @Nullable
    private final ConstrainedDimension maxWidth;
    @Nullable
    private final ConstrainedDimension maxHeight;

    public ConstrainedSize(@NonNull String width, @NonNull String height,
                           @Nullable String minWidth, @Nullable String minHeight,
                           @Nullable String maxWidth, @Nullable String maxHeight) {
        super(width, height);
        this.minWidth = ConstrainedDimension.of(minWidth);
        this.minHeight = ConstrainedDimension.of(minHeight);
        this.maxWidth = ConstrainedDimension.of(maxWidth);
        this.maxHeight = ConstrainedDimension.of(maxHeight);
    }

    @NonNull
    public static ConstrainedSize fromJson(@NonNull JsonMap json) throws JsonException {
        String width = json.opt("width").coerceString();
        String height = json.opt("height").coerceString();
        if (width == null || height == null) {
            throw new JsonException("Size requires both width and height!");
        }

        String minWidth = json.opt("min_width").coerceString();
        String minHeight = json.opt("min_height").coerceString();
        String maxWidth = json.opt("max_width").coerceString();
        String maxHeight = json.opt("max_height").coerceString();

        return new ConstrainedSize(width, height, minWidth, minHeight, maxWidth, maxHeight);
    }

    @NonNull
    @Override
    public String toString() {
        return "ConstrainedSize { " +
            "width=" + getWidth() + ", height=" + getHeight() +
            ", minWidth=" + getMinWidth() + ", minHeight=" + getMinHeight() +
            ", maxWidth=" + getMaxWidth() + ", maxHeight=" + getMaxHeight() +
            " }";
    }

    @Nullable
    public ConstrainedDimension getMinWidth() {
        return minWidth;
    }

    @Nullable
    public ConstrainedDimension getMinHeight() {
        return minHeight;
    }

    @Nullable
    public ConstrainedDimension getMaxWidth() {
        return maxWidth;
    }

    @Nullable
    public ConstrainedDimension getMaxHeight() {
        return maxHeight;
    }

    public enum ConstrainedDimensionType {
        PERCENT,
        ABSOLUTE
    }

    public abstract static class ConstrainedDimension {
        @NonNull
        protected final String value;
        @NonNull
        private final ConstrainedDimensionType type;

        ConstrainedDimension(@NonNull String value, @NonNull ConstrainedDimensionType type) {
            this.value = value;
            this.type = type;
        }

        @Nullable
        public static ConstrainedDimension of(@Nullable String value) {
            if (value == null) {
                return null;
            } else if (PercentUtils.isPercent(value)) {
                return new PercentConstrainedDimension(value);
            } else {
                return new AbsoluteConstrainedDimension(value);
            }
        }

        public abstract float getFloat();

        public abstract int getInt();

        @NonNull
        public ConstrainedDimensionType getType() {
            return type;
        }

        public boolean isPercent() {
            return type == PERCENT;
        }

        public boolean isAbsolute() {
            return type == ABSOLUTE;
        }
    }

    public static class PercentConstrainedDimension extends ConstrainedDimension {
        PercentConstrainedDimension(@NonNull String value) {
            super(value, PERCENT);
        }

        @Override
        public float getFloat() {
            return PercentUtils.parse(value);
        }

        @Override
        public int getInt() {
            return (int) getFloat();
        }

        @NonNull
        @Override
        public String toString() {
            return (int) (getFloat() * 100) + "%";
        }
    }

    public static class AbsoluteConstrainedDimension extends ConstrainedDimension {
        AbsoluteConstrainedDimension(@NonNull String value) {
            super(value, ABSOLUTE);
        }

        @Override
        public float getFloat() {
            return Float.parseFloat(value);
        }

        @Override
        public int getInt() {
            return Integer.parseInt(value);
        }

        @NonNull
        @Override
        public String toString() {
            return getInt() + "dp";
        }
    }
}

File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/property/ConstrainedSize.kt
/* Copyright Airship and Contributors */
package com.urbanairship.android.layout.property

import com.urbanairship.android.layout.util.PercentUtils
import com.urbanairship.json.JsonException
import com.urbanairship.json.JsonValue

public class ConstrainedSize public constructor(
    width: String,
    height: String,
    minWidth: String?,
    minHeight: String?,
    maxWidth: String?,
    maxHeight: String?
) : Size(width, height) {

    @JvmField
    public val minWidth: ConstrainedDimension?
    @JvmField
    public val minHeight: ConstrainedDimension?
    @JvmField
    public val maxWidth: ConstrainedDimension?
    @JvmField
    public val maxHeight: ConstrainedDimension?

    init {
        this.minWidth = ConstrainedDimension.Companion.of(minWidth)
        this.minHeight = ConstrainedDimension.Companion.of(minHeight)
        this.maxWidth = ConstrainedDimension.Companion.of(maxWidth)
        this.maxHeight = ConstrainedDimension.Companion.of(maxHeight)
    }

    override fun toString(): String {
        return "ConstrainedSize { " +
                "width=$width, height=$height, " +
                "minWidth=$minWidth, minHeight=$minHeight, " +
                "maxWidth=$maxWidth, maxHeight=$maxHeight }"
    }

    public enum class ConstrainedDimensionType {
        PERCENT, ABSOLUTE
    }

    public abstract class ConstrainedDimension internal constructor(
        protected val value: String,
        @JvmField public val type: ConstrainedDimensionType
    ) {

        public abstract fun getFloat(): Float

        public abstract fun getInt(): Int

        public fun isPercent(): Boolean {
            return type == ConstrainedDimensionType.PERCENT
        }

        public fun isAbsolute(): Boolean {
            return type == ConstrainedDimensionType.ABSOLUTE
        }

        public companion object {

            public fun of(value: String?): ConstrainedDimension? {
                val content = value ?: return null

                return if (PercentUtils.isPercent(content)) {
                    PercentConstrainedDimension(content)
                } else {
                    AbsoluteConstrainedDimension(content)
                }
            }
        }
    }

    public class PercentConstrainedDimension internal constructor(
        value: String
    ) : ConstrainedDimension(value, ConstrainedDimensionType.PERCENT) {

        override fun getFloat(): Float {
            return PercentUtils.parse(value)
        }

        override fun getInt(): Int {
            return getFloat().toInt()
        }

        override fun toString(): String {
            return (getFloat() * 100).toInt().toString() + "%"
        }
    }

    public class AbsoluteConstrainedDimension internal constructor(
        value: String
    ) : ConstrainedDimension(value, ConstrainedDimensionType.ABSOLUTE) {

        override fun getFloat(): Float {
            return value.toFloat()
        }

        override fun getInt(): Int {
            return value.toInt()
        }

        override fun toString(): String {
            return getInt().toString() + "dp"
        }
    }

    public companion object {

        private const val KEY_WIDTH = "width"
        private const val KEY_HEIGHT = "height"
        private const val KEY_MIN_WIDTH = "min_width"
        private const val KEY_MIN_HEIGHT = "min_height"
        private const val KEY_MAX_WIDTH = "max_width"
        private const val KEY_MAX_HEIGHT = "max_height"

        @Throws(JsonException::class)
        public fun fromJson(json: JsonValue): ConstrainedSize {
            val content = json.requireMap()

            val width = content[KEY_WIDTH]?.coerceString()
            val height = content[KEY_HEIGHT]?.coerceString()
            if (width == null || height == null) {
                throw JsonException("Size requires both width and height!")
            }

            return ConstrainedSize(
                width = width,
                height = height,
                minWidth = content[KEY_MIN_WIDTH]?.coerceString(),
                minHeight = content[KEY_MIN_HEIGHT]?.coerceString(),
                maxWidth = content[KEY_MAX_WIDTH]?.coerceString(),
                maxHeight = content[KEY_MAX_HEIGHT]?.coerceString()
            )
        }
    }
}
